pkg/asset/agent/manifests/nmstateconfig.go

```diff
@@ -336,3 +339,33 @@ func buildMacInterfaceMap(nmStateConfig aiv1beta1.NMStateConfig) models.MacInter
 	}
 	return macInterfaceMap
 }
+
+func checkHostCount(installConfig *types.InstallConfig, agentConfig *agentconfig.AgentConfig) {
+	numRequiredMasters, numRequiredWorkers := agent.GetReplicaCount(installConfig)
+
+	// Check that the number of replicas matches the configured hosts
+	numMasters := int64(0)
+	numWorkers := int64(0)
+	for _, host := range agentConfig.Config.Hosts {
+		switch host.Role {
+		case "master":
+			numMasters++
+		case "worker":
+			numWorkers++
+		case "":
+			// If not defined, the roles will be matched to replicas
+			if numMasters < numRequiredMasters {
+				numMasters++
+			} else if numWorkers < numRequiredWorkers {
+				numWorkers++
+			}
+		}
+	}
```

Comment:
This will fail spuriously in a case like:

```
- role: ""
- role: ""
- role: master
- role: master
- role: master
```

We need to iterate over the hosts twice, once to count the explicit roles and once for the implicit ones.

Reply:
I think that the original validation was taken from the https://github.com/openshift/installer/pull/5205, where the very same test scenario was addressed by having the hosts [sorted by role](https://github.com/openshift/installer/blob/063245f6b4cdb387eeff331e2b6f7e5ea47038fb/pkg/types/baremetal/defaults/platform.go#L140-L143) before applying the validation.

Reply:
Thanks @andfasano, yes it was based on the baremetal validation. As this list will be sorted I think we can use the one loop as it currently has and not need to iterate over the hosts a 2nd time.